### ISSUE TITLE
Alternative & new OSS-TTS-engines

### ISSUE THREAD

**Issue by Olf0** [rinigus/pure-maps#627]
### New contenders: RHVoice & eSpeakNG

I remembered the side-tracked discussion about [alternative and maintained OSS-TTS-engines](https://github.com/rinigus/pure-maps/issues/34#issuecomment-443525638), when I came across these two TTS engines at F-Droid ([[1]](https://f-droid.org/en/packages/com.github.olga_yakovleva.rhvoice.android/), [[2]](https://f-droid.org/en/packages/com.reecedunn.espeak/)):
* https://github.com/RHVoice/RHVoice#readme
* https://github.com/espeak-ng/espeak-ng#readme

Both are maintained, but eSpeakNG might output a lower quality voice, as most improvements relative to the original eSpeak ([[1]](https://sourceforge.net/p/espeak/code/HEAD/tree/), [[2]](https://github.com/eeejay/espeak/issues/9#issue-1347022437), [[3]](https://github.com/Olf0/eSpeak/issues/1#issue-1347100972)) did not address the voice engine proper.  OTOH, eSpeakNG now also has [Python bindings contributed](https://github.com/gooofy/py-espeak-ng), and the [eSpeakNG-based Mimic3 listening samples are fine](https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mimic-tts/).  eSpeakNG is well documented: [[1]](https://github.com/espeak-ng/espeak-ng#readme) & [[2]](https://github.com/espeak-ng/espeak-ng/tree/master/docs) 

In comparison to eSpeak (and maybe also eSpeakNG), [RHVoice](https://github.com/RHVoice/RHVoice#readme) seems to provide a higher-quality voice-synthesis and a set of languages which often lack good quality speech-synthesis.  RHVoice is [documented in multiple languages](https://github.com/RHVoice/RHVoice/tree/master/doc).

### flite

Then I pursued to look up descendents of CMU's flite (Carnegie Mellon University's Festival-lite vox (voice encoder)), so I first started looking for flite proper: [latest source code]() v2.1 - v2.3+, [original source code](http://festvox.org/flite/download.html) ([alternative site](http://cmuflite.org/download.html)) v1.0 - v2.1.0, [research context](http://festvox.org/), [original webpage with slide deck and scientific paper](http://cmuflite.org/) ([alternative site](http://festvox.org/flite/), direct link to paper as [HTML pages](https://www.cs.cmu.edu/~awb/papers/ISCA01/flite/flite.html) and [Postscript file](http://www.cs.cmu.edu/~awb/papers/ISCA01/flite.ps)) Side note: Interesting to still see [occasional commits](https://github.com/festvox/flite/commits/master) (as of 2022-08-23, the latest one on 2022-05-16) and releases [2.2 on 2020-08-13](https://github.com/festvox/flite/releases) and [2.3 in "March 2022"](https://github.com/festvox/flite#news), but [untagged](https://github.com/festvox/flite/tags), hence better consider the [*master* branch as of 2022-06-15](https://github.com/festvox/flite/commit/6c9f20dc915b17f5619340069889db0aa007fcdc) as "flite 2.3+" release) for the original flite.

### mimic 1, 2 & 3

Then I looked at the well known flite-based mimic, of which we knew that its first incarnation had ceased development with its [release v1.3.0.1](https://github.com/MycroftAI/mimic1/releases) (also the [state of the *master* branch](https://github.com/MycroftAI/mimic1/tree/master) since 2020-03-06) with a [few additional bug-fixes in the *development* branch](https://github.com/MycroftAI/mimic1/commits/development), works well (solely for English) and is well documented ([[1]](https://github.com/MycroftAI/mimic1/tree/development#readme) & [[2]](https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mimic-tts/mimic-overview)).

[@rinigus had hopes for mimic2](https://github.com/rinigus/pure-maps/issues/34#issuecomment-443529363) ([see also](https://mimic.mycroft.ai/)), which [died quickly in 2020](https://github.com/MycroftAI/mimic2), despite [having the cool ability to deploy ones own voice](https://github.com/MycroftAI/mimic-recording-studio).
Furthermore, it was ["designed to run in the cloud"](https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mimic-tts/#mimic-2)], which likely means that it uses a lot of resources when the server and client component are running on a single machine.
Mimic2 is also very well documented: [[1]](https://github.com/MycroftAI/mimic2#readme) & [[2]](https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mimic-tts/mimic-2)

[Mimic-3 is now Mycrofts's new focus](https://mycroft.ai/mimic-3/) and it seems to [be developed well](https://github.com/MycroftAI/mimic3), but only had a [single, proper release yet](https://github.com/MycroftAI/mimic3/releases) (v0.2.3) and uses [`libespeak-ng1`](https://github.com/MycroftAI/mimic3#mycroft-tts-plugin).
It is also supported by the cool [Mimic Recording Studio](https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mimic-tts/#mimic-recording-studio) to record one's own voice.
Mimic-3 is nicely documented, too: [[1]](https://github.com/MycroftAI/mimic3#readme) and [[2]](https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mimic-tts/mimic-3)  Still it needs to be analysed which functions mimic-3 provides over a [direct use of `libespeak-ng`](https://github.com/espeak-ng/espeak-ng/blob/master/docs/integration.md#using-espeak-ng-as-a-library) and evaluated if these are worth the additional dependency (technically and WRT sustainability).

Interestingly [Mycroft's top-level mimic documentation-page](https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mimic-tts/) provides listening samples on which the `libespeak-ng1`-based mimic3 output sounds quite well.

### FreeTTS

[FreeTTS](https://sourceforge.net/projects/freetts/) is also flite-based, but written in Java.  Furthermore [its latest release is v1.2.2 on 2009-03-09](https://sourceforge.net/projects/freetts/files/) and [its la(te)st commit to SVN-trunk happened on 2012-05-08](https://sourceforge.net/p/freetts/code/HEAD/tree/).  Thus not worth to pursue.

### NanoTTS, a command-line front-end for PicoTTS

[NanoTTS](https://github.com/gmn/nanotts) [ceased development in 2019](https://github.com/gmn/nanotts/commits/master), while [the la(te)st commit to PicoTTS happened on 2018-02-14](https://android.googlesource.com/platform/external/svox/+/master).  As mimic1, NanoTTS and PicoTTS are clearly EOLed, but working fine (in my experience), plus support more languages than mimic1.

# TL;DR

[RHVoice](https://github.com/RHVoice/RHVoice) ([documentation](https://github.com/RHVoice/RHVoice/tree/master/doc)) seems to be worth being evaluated for integration in Pure Maps and so does [eSpeakNG](https://github.com/espeak-ng/espeak-ng/) / [`libespeak-ng`](https://github.com/espeak-ng/espeak-ng/blob/master/docs/integration.md#using-espeak-ng-as-a-library) (documentation: [[1]](https://github.com/espeak-ng/espeak-ng#readme) & [[2]](https://github.com/espeak-ng/espeak-ng/tree/master/docs)) and / or [mimic-3](https://github.com/MycroftAI/mimic3) (documentation [[1]](https://mycroft.ai/mimic-3/), [[2]](https://github.com/MycroftAI/mimic3#readme), [[3]](https://mycroft-ai.gitbook.io/docs/mycroft-technologies/mimic-tts/mimic-3)) in order to provide maintained and improved TTS synthesis compared to the extant choices mimic1 and NanoTTS (includes PicoTTS).  As these legacy components are working fine (well, mimic1 never for me, but for many others), there is no currently need to rush the evaluation and potential adaption of RHVoice and / or eSpeakNG / `libespeak-ng` / mimic-3.

**Comment by rinigus** [comment]
Thank you very much for this study and summary!

**Comment by JacekJagosz** [comment]
I would vote using Mimic 3, it sounds the best to me, and supports a decent number of languages. And eSpeakNG is a very decent choice too, seeing how many languages it supports.
Not sure if RHVoice should be the first choice, as it doesn't support a lot of languages currently supported by Pure Maps. But it is just my opinion, awesome writeup by Olf0

**Comment by Olf0** [comment]
> I would vote using Mimic 3, […] And eSpeakNG is a very decent choice too, seeing how many languages it supports.

But if mimic3 is not much more that eSpeakNG (still needs to be checked), using `libespeak-ng` directly may be better, slenderer, easier socially & technically etc.

> Not sure if RHVoice should be the first choice, as it doesn't support a lot of languages currently supported by Pure Maps.

It supports Polish. :wink:

More seriously: It all is about the quality of the synthesised speech in relation to the resources (CPU, RAM, I/O, mass storage) used.
You may check, if RHVoice runs under AlienDalvik (or install it natively, if you have a Android device; it is [available from F-Droid](https://f-droid.org/en/packages/com.github.olga_yakovleva.rhvoice.android/)) and compare its quality with the one of mimic1, PicoTTS (NanoTTS) etc. (I usually also have the Google TTS engine installed on AlienDalvik).

**Comment by rinigus** [comment]
One more suggested by piggz: https://github.com/coqui-ai/TTS . looks to be distributed via pip, at least on PC
